src/main/groovy/lang/IntRange.java

```diff
@@ -355,40 +360,54 @@ public boolean containsAll(Collection other) {
         return super.containsAll(other);
     }
 
+    @Override
     public void step(int step, Closure closure) {
-        if (step == 0) {
-            if (!getFrom().equals(getTo())) {
+        step((Number) step, closure);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void step(Number step, Closure closure) {
+        Long from = getFrom().longValue();
+        Long to = getTo().longValue();
+
+        if (compareEqual(step, 0)) {
+            if (compareNotEqual(from, to)) {
                 throw new GroovyRuntimeException("Infinite loop detected due to step size of 0");
             } else {
                 return; // from == to and step == 0, nothing to do, so return
             }
         }
 
-        if (isReverse()) {
-            step = -step;
+        boolean isAscending = !isReverse();
+        if (compareLessThan(step, 0)) {
+            step = multiply(step, -1);
+            isAscending = !isAscending;
         }
-        if (step > 0) {
-            int value = getFrom();
-            while (value <= getTo()) {
-                closure.call(Integer.valueOf(value));
-                if ((0L + value + step) >= Integer.MAX_VALUE) {
-                    break;
-                }
-                value = value + step;
+
+        if (isAscending) {
+            for (Long value = from; value <= to; value = (Long) plus(value, step)) {
```

Comment:
It could even be `Double.POSITIVE_INFINITY` or `Double.NaN` or `Double.MIN_VALUE` and probably none would work "correctly".